### PR TITLE
Make it possible to get a StackReference output promptly

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
A customer has run into a situation where they are exporting information from one stack and want to read it in another stack.  hte information in this case relates to their Vpc and specifically has the `vpcId` and the `publicSubnetIds`.  

In their other stack, they want to read this data in and then hydrate an `awsx.ec2.Vpc` using `.fromExistingIds`.  However, this currently doesn't work because `fromExistingIds` wants the data (specifically the arrays) to be prompt.  That's because it uses those arrays to actually create the Subnet resources promptly so they can be used immediately once the Vpc is completed. 

Because StackReference values are 'outputs' this doesn't work.  The typings don't match. A nd even if they did, there's no way for the Vpc to promptly iterate teh subnetIds to create the Subnet resources.

To alleviate this problem, we are introducing `StackRef.getOutputSync` which blocks until the value is actually available.  With this the second stack can now get the real array of subnetIds and can hydrate the awsx.ec2.Vpc from tehm promptly.

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Note: i haven't tested this yet.  Going to do so now.

**Comment by ellismg** [comment]
My personal opinion: It would be helpful if we included either in the PR (or ideally the commit message) why we are making this change?  It feels a little weird to me that we are just dropping the `Output`'ness here because in addition to making the value prompt we are also losing the dependency information.  I think it would be helpful if we wrote down why we are making this trade-off (I assume it is because we want folks to be able to do control flow on these values, but would be nice to have some more details about the tradeoffs here).

**Comment by pgavlin** [comment]
My major concerns relate to the fact that the synchronously-available value drops the dependency and secret-ness information associated with the stack output. The former is probably not much of an issue for `StackReference`--the dependency information is really only important for deletes, so the only place that I think this hurts is when the name of the referenced stack is an `Input` and we lose transitive dependency info--but the latter seems like it could be pretty bad.

**Comment by CyrusNajmabadi** [comment]
> My personal opinion: It would be helpful if we included either in the PR (or ideally the commit message) why we are making this change?

Totally, and that will def be added.  This was more to just get some initial eyes.  We've been having a convo about this on slack, and i'll def update things accordingly.

**Comment by ellismg** [comment]
> but the latter seems like it could be pretty bad.

I agree with this.  It is unfortunate that because of our model today if any stack output is a secret, then the entire Outputs `Output<T>` is treated as a secret (because only Output<T> can be a secret), if that wasn't the case, it seems like you could at least say that this Sync version throws if the value it would return would be secret and you'd have sort of OKish UX.  But as is having a single secret output would mean that fetching /any/ output would trigger that exception.

It might still be good to start with throwing on secrets, FWIW.

**Comment by CyrusNajmabadi** [comment]
> It might still be good to start with throwing on secrets, FWIW.

Yes, i'm going to start with that at least.

**Comment by CyrusNajmabadi** [comment]
OP updated.

File: sdk/nodejs/stackReference.ts [hunk 1]
```diff
@@ -56,6 +57,25 @@ export class StackReference extends CustomResource {
...
+     * This operation is not supported (and will throw) if any exported values of the StackReference
+     * are secrets.
+     *
+     * @param name The name of the stack output to fetch.
+     */
+    public getOutputSync(name: string): any {
+        const out = this.getOutput(name);
+        const isSecret = promiseResult(out.isSecret);
+        if (isSecret) {
+            throw new Error("Cannot call [getOutputSync] on a StackReference containing secrets.  Only [getOutput] is supported.");
```
Comment (pgavlin) [inline sdk/nodejs/stackReference.ts]: Might be clearer to say "Cannot call [getOutputSync] if the referenced stack has secret outputs. Use [getOutput] instead."

Comment (CyrusNajmabadi) [reply sdk/nodejs/stackReference.ts]: Sure!

**Comment by CyrusNajmabadi** [comment]
Ok.  So this all passed and worked when i manually tested it for the Vpc scenario.  @pgavlin @ellismg @lukehoban give me your plus/minus on going forward with this.  Are the pros worth it?  Or do you think the cons outweigh things.

Personally, i think it's reasonable to go this path.  The facility is isolated to just this ability, and the user use case is clear and reasonable.  I dont think this will be open for abuse or will really lead to bad scenarios.

**Comment by ellismg** [comment]
> Personally, i think it's reasonable to go this path. The facility is isolated to just this ability, and the user use case is clear and reasonable. I dont think this will be open for abuse or will really lead to bad scenarios.

Should we also be doing this in Python?

**Comment by pgavlin** [comment]
I do worry that the all-or-nothing nature of secret stack outputs could make this difficult to use in practice, but I don't know that we should block on that. @ellismg have you done any thinking about how we could address the secret output issue? I wonder if the provider could return a map from property name to secret-ness that we could use in `getOutput`/`getOutputSync`.

**Comment by ellismg** [comment]
> I wonder if the provider could return a map from property name to secret-ness that we could use in getOutput/getOutputSync.

Yes, that would certainly be a strategy.  I had not fully thought about it, but we do have https://github.com/pulumi/pulumi/issues/2744 tracking the issue.

**Comment by CyrusNajmabadi** [comment]
From talking to Pat online, it sounded like we were good with this.  It may be necessary in the future to do things in a more fine-grained fashion if we find lots of users running into the limitation with secrets.  However, this work doesn't conflict with that.

**Pull request merged**